larray/core/session.py

```diff
@@ -1475,6 +1478,164 @@ def display(k, v, is_metadata=False):
+    ...     mortality_rate = ArrayDef((AGE, GENDER))
+
+    >>> def run_model(variant_name, first_proj_year):
+    ...     # create an instance of the ModelVariables class
+    ...     m = ModelVariables()
+    ...     # ==== setup variables ====
+    ...     # set scalars
+    ...     m.variant_name = variant_name
+    ...     m.first_proj_year = first_proj_year
+    ...     # set axes
+    ...     m.obs_years = Axis(f'time={m.FIRST_OBS_YEAR}..{m.first_proj_year-1}')
+    ...     m.proj_years = Axis(f'time={m.first_proj_year}..{m.LAST_PROJ_YEAR}')
+    ...     # set arrays
+    ...     m.population_obs = zeros((m.AGE, m.GENDER, m.obs_years))
+    ...     m.population_proj = zeros((m.AGE, m.GENDER, m.proj_years))
+    ...     m.mortality_rate = zeros((m.AGE, m.GENDER))
+    ...     # ==== model ====
+    ...     # some code here
+    ...     # ...
+    ...     # ==== output ====
+    ...     # save all variables in an HDF5 file
+    ...     m.save('{variant_name}.h5', display=True)
+
+    Content of file 'main.py'
+
+    >>> run_model('projection_2020_2070', first_proj_year=2020)
+    dumping FIRST_OBS_YEAR ... done
+    dumping LAST_PROJ_YEAR ... done
+    dumping AGE ... done
+    dumping GENDER ... done
+    dumping G_CHILDREN ... done
+    dumping G_ADULTS ... done
+    dumping variant_name ... done
+    dumping first_proj_year ... done
+    dumping obs_years ... done
+    dumping proj_years ... done
+    dumping population_obs ... done
+    dumping population_proj ... done
+    dumping mortality_rate ... done
+    """
+    def __init__(self, *args, **kwargs):
+        _cls_attrs = {key: value for key, value in vars(self.__class__).items() if not key.startswith('_')}
+        object.__setattr__(self, '_cls_attrs', _cls_attrs)
+
+        constant_cls_attrs = {key: value for key, value in _cls_attrs.items()
+                              if not isinstance(value, (type, ArrayDef))}
+
+        if len(args) == 1:
+            # Note: using the two lines below we may loose the order in which variables have been declared
+            # but can we avoid it?
+            Session.__init__(self, *args, **kwargs)
+            self.add(**constant_cls_attrs)
+        else:
+            # Note: same as above
+            kwargs.update(constant_cls_attrs)
+            Session.__init__(self, *args, **kwargs)
+
+    def load(self, fname, names=None, engine='auto', display=False, **kwargs):
+        super().load(fname, names, engine, display, **kwargs)
+        for key in set(self._cls_attrs.keys()) - set(self.keys()):
+            warnings.warn(f"The variable {key} is declared in the {self.__class__.__name__} "
+                          f"class definition but has not found in file {fname}.")
+        for key in set(self.keys()) - set(self._cls_attrs.keys()):
+            warnings.warn(f"The variable {key} has been found in file {fname} but is not declared "
+                          f"in the  {self.__class__.__name__} class definition.")
```

Comment:
I am **unsure** this warning should exist either. It seems reasonable to me to instantiate a session with a file containing more variables than necessary. LIAM2 accepts this kind of things (if I transpose the concepts) and users have found this feature quite useful. The trick is that, in that case LIAM2 only loads the declared stuff. **Maybe** we should do this here too?

Reply:
> The trick is that, in that case LIAM2 only loads the declared stuff.
So LIAM2 only loads the declared stuff but does not print any warnings if there are undeclared stuff in the file?

Reply:
Yes. The spirit is that you declare the fields (LIAM2 vocabulary) you need in your model and can used whatever file which provides those fields. This makes it possible to use the same input data files for several variants, some of which need extra fields.